grpc/src/tx.rs

```diff
@@ -0,0 +1,602 @@
+        account.sequence += 1;
+
+        Ok((resp.txhash, tx_sequence))
+    }
+
+    async fn confirm_tx(&self, hash: Hash, sequence: u64) -> Result<TxInfo> {
+        let mut interval = Interval::new(Duration::from_millis(500)).await;
+
+        loop {
+            let tx_status = self.client.tx_status(hash).await?;
+            match tx_status.status {
+                TxStatus::Pending => interval.tick().await,
+                TxStatus::Unknown => return Err(Error::TxNotFound(hash)),
+                TxStatus::Evicted => {
+                    // node will treat this transaction like if it never happened, so
+                    // we need to revert the account's sequence to the one of evicted tx.
+                    // all transactions that were already submitted after this one
+                    // will fail due to incorrect sequence number.
+                    let mut acc = self.account.lock().await;
+                    acc.sequence = sequence;
+                    return Err(Error::TxEvicted(hash));
+                }
+                TxStatus::Committed => {
+                    if tx_status.execution_code == ErrorCode::Success {
+                        return Ok(TxInfo {
+                            hash,
+                            height: tx_status.height,
+                        });
+                    } else {
+                        return Err(Error::TxExecutionFailed(
+                            hash,
+                            tx_status.execution_code,
+                            tx_status.error,
+                        ));
+                    }
+                }
+            }
+        }
+    }
+
+    fn maybe_update_gas_price(&self, err: &Error, cfg: TxConfig) -> Result<bool> {
+        let Error::TxBroadcastFailed(_, code, error, gas_limit) = err else {
+            return Ok(false);
+        };
+        // nothing to update if we didn't use our gas internal gas price
+        if cfg.gas_price.is_some() {
+            return Ok(false);
+        }
+        if *code != ErrorCode::InsufficientFee {
+            return Ok(false);
+        }
+
+        let Some((got_fee, want_fee)) = parse_insufficient_gas_err(error) else {
+            return Err(Error::UpdatingGasPriceFailed(format!(
+                "Couldn't parse required fee from error: {err}"
+            )));
+        };
+        if want_fee == 0.0 {
+            return Err(Error::UpdatingGasPriceFailed(format!(
+                "Wanted fee is 0: {err}"
+            )));
+        }
+
+        let new_gas_price = if self.gas_price() == 0.0 || got_fee == 0.0 {
+            if *gas_limit == 0 {
+                return Err(Error::UpdatingGasPriceFailed(format!(
+                    "Cannot update gas price when gas price and limit are 0: {err}"
+                )));
+            }
+            want_fee / *gas_limit as f64
+        } else {
+            want_fee / got_fee * self.gas_price()
+        };
+
+        self.set_gas_price(new_gas_price.ceil());
```

Comment:
Shouldn't the gas price be per transaction? Why do we save it? If gas goes 2x up and after some time down 50%, the user will pay the 2x price.

Reply:
done